src/goto-instrument/code_contracts.cpp

```diff
@@ -1542,11 +1542,38 @@ goto_programt assigns_clauset::havoc_code(
   goto_programt havoc_statements;
   for(assigns_clause_targett *target : targets)
   {
+    //    if(target != NULL) goto x;
+    //      havoc_statements
+    // z: skip
+
+    // create the z label
+    goto_programt tmp_z;
+    goto_programt::targett z = tmp_z.add(goto_programt::make_skip(location));
+
+    // TODO: generalize this. Currently only supports pointers
+    // See GitHub issue #6087 for further details
+    if(target->target_type == assigns_clause_targett::target_type::Scalar)
+    {
+      exprt condition = equal_exprt(
+        target->get_direct_pointer(),
+        null_pointer_exprt(
+          to_pointer_type(target->get_direct_pointer().type())));
```

Comment:
This condition would not work for invalid pointers though. I would suggest checking validity of pointers, not just `NULL`-ness.

One way to check validity in this context would be to check `__CPROVER_w_ok(target, 0)`, something like:
```c
      exprt condition = w_ok_exprt(target->get_direct_pointer(), from_integer(0, int_type));
```
(not super sure about the syntax, but I hope you get the idea)

This is inspired by a similar validity check being done within the implementation of `free`:
https://github.com/diffblue/cbmc/blob/48893287099cb5780302fe9dc415eb6888354fd6/src/ansi-c/library/stdlib.c#L250-L252

Reply:
We tested 4 things, and w_ok doesn't work as expected on one of them:
- `!__CPROVER_w_ok(NULL, 0)` returns `true`
- `!__CPROVER_w_ok(ptr, 0)` returns `true` on an invalid memory location, like `ptr = 0x12345678`
- `!__CPROVER_w_ok(ptr, 0)` returns `false` on a valid (non-`NULL` result from `malloc`) memory location
- `!__CPROVER_w_ok(ptr, 0)` returns `false` on an uninitialized `ptr`, which is unexpected